### DISCUSSION TITLE
Using an `FnMut` in a button closure?

### DISCUSSION THREAD

**Discussion by ezntek** [gyscos/cursive#733] [Q&A]
I need to have a button that mutates some data, under the context of a _state manager_. I wrote a simple state manager that works kind of like a state machine with changes in state happening through navigating the UI, and some functions related to mutating the data requires a mutable reference to self, which requires a mutable variable and a mutable reference, which cannot be obtained inside a plain `Fn`. Is there a way around this:

Note: I did try wrapping it in an `Rc` of `Cell`, it did work until I encountered `closure may outlive the current function, but it borrows \`state manager variable name\`, which is owned by the current function.` Using `move ` would work, but one would need to make a different clone of the `Rc` per closure, which is kind of clunky.

**Comment by gyscos** [comment]
Hi, and thanks for the question!

Indeed, a button currently stores a `Fn`, because in some cases the callback could be called recursively (for example if the callback trigger an event that results in the same callback), and `FnMut` cannot support that.

If you don't care about the recursive situation, you can wrap a `FnMut` into a `Fn` through a `Rc` and `Cell` as you mentioned (or more precisely, a `Mutex` now that cursive moved to be `Sync`). I added a macro to help do it for you: [`cursive::immut1!`](https://docs.rs/cursive/latest/cursive/macro.immut1.html). If you're curious you can see [how it's implemented](https://docs.rs/cursive_core/0.4.6/src/cursive_core/utils/immutify.rs.html#76), but it's really just using a `Mutex` to get mutable access to the inner closure, skipping if it's already locked (recursive call).
